wordpress-mobile/WordPress-iOS
# Editor: Additional error messages for media upload failures

Fixes https://github.com/wordpress-mobile/WordPress-iOS/issues/14400

## Description

When an image fails to load via the image block in the editor (for any reason other than there being no Internet connection) a progress indicator currently remains indefinitely on screen. The user will only see an error if they tap on the image (in which case an error appears in the Retry action sheet) or if they save the post as a draft.

## Proposed changes

When an upload's thumbnail is ready, we don't currently check whether the upload failed, we only [check whether the app is connected to the Internet](https://github.com/wordpress-mobile/WordPress-iOS/blob/1203aa44027b4b1b76dffcb46c6973715bcc4221/WordPress/Classes/ViewRelated/Gutenberg/GutenbergMediaInserterHelper.swift#L240) (as per https://github.com/wordpress-mobile/WordPress-iOS/pull/14140). That leads to a situation where uploads that fail (for reasons other than no Internet connection) are assigned an "uploading" state via [this line of code](https://github.com/wordpress-mobile/WordPress-iOS/blob/8ebc719bf1e4416dad99b60c564cc1f43af1ed0f/WordPress/Classes/ViewRelated/Gutenberg/GutenbergMediaInserterHelper.swift#L245). This sends a signal back to Gutenberg and prevents an error message for the failure from being displayed correctly. 

With this PR, a check for whether an upload failed has been added to guard against that in the `thumbnailReady` state [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/8ebc719bf1e4416dad99b60c564cc1f43af1ed0f/WordPress/Classes/ViewRelated/Gutenberg/GutenbergMediaInserterHelper.swift#L241). I chose to use `media.remoteStatus == .failed` for the check as it matches the pattern used in [the switch statement here](https://github.com/wordpress-mobile/WordPress-iOS/blob/cb0c17f175d6db33ce928f36156bc70907b17889/WordPress/Classes/ViewRelated/Media/MediaLibraryViewController.swift#L414) for uploads to the Media Library.

## Screenshots

You'll see that, prior to this change, an upload failed "silently", with no error message to indicate that it was not successful. The changes in this PR address that issue so that the error message now displays as expected following the failure:

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/154478837-93714454-6b71-45da-b247-81f6f4a27eeb.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/154477890-029d2523-9215-49ec-bcc4-957d4e768095.png" width="100%"> |

## Testing

_Prerequisite: A WordPress.com Simple site with a media library that's been filled up to the site's storage limit. The credentials for an existing test site that matches that criteria be found here: paaHJt-17e-p2#comment-2395_ 

Steps to test:

1. Start a new post.
2. Add an image block.
3. Select an image to upload.
4. Desired result: The image placeholder appears with an error message. 
5. Prior to this change: The error message did not display, instead there was a permanent placeholder/progress indicator.

## Regression Notes
1. Potential unintended areas of impact

The code area that's been tweaked with this change relates to the way an image's thumbnail appears when it's uploaded in the editor. It may therefore unintentionally impact the way the thumbnail appears.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manually tested uploading an image to ensure the thumbnail displays as expected, both when the upload fails and when it's successful.

3. What automated tests I added (or what prevented me from doing so)

No automated tests have been added as this is a relatively small fix for the UI, I didn't feel an automated test was necessary.

<hr />

PR submission checklist:
